The possibility to specify a path to the settings file

kudashevs opened the pull request:
Hello again :wave:,

This is a draft PR of the first step of [a proposition](https://github.com/spatie/laravel-ignition/discussions/48) to the `laravel-ignition`.

This is a new component `FileConfigManager` which is implemented through a new `ConfigManager` interface. The interface is a little bit over-engineered with the `getSource` method, however, I've found it useful for testing purposes and it could be useful in the future (but it could be removed, though). So, the new `FileConfigManager` gets all the responsibility to manage the settings file:

- it accepts a path through the constructor and validates it
- if the path is not provided it tries to find an appropriate path from the environment (was taken from DefaultConfigFinder)
- it can load the options from a file
- it can save the options to the file
- it can create the file (and this seems the most important because DefaultConfigFinder didn't do that, so DefaultConfigFinderTest didn't work properly for me if I didn't have a created settings file)
- it is covered with tests in some weak points

AlexVanderbist commented on src/Config/FileConfigManager.php:
```diff
@@ -0,0 +1,142 @@
+
+class FileConfigManager implements ConfigManager
+{
+    private const SETTINGS_FILE_NAME = '.ignition.json';
```
Maybe we can rename this to `CONFIG_FILE_NAME` to keep it in line with the `ConfigManager`, etc...

```suggestion
    private const CONFIG_FILE_NAME = '.ignition.json';
```

kudashevs replied:
Hello,

I kind of agree, but to me, the `config/ignition.php` is a config file and `.ignition.json` is a settings file. One of the reasons for this separation is the Laravel Ignition button naming (we have a Save settings button there). So, not to confuse users I tried to differentiate these two concepts by introducing this term `settings file`. That's why I would like to stick to this name `settings file` and use it for this file if you don't mind.

However, if you don't like this differentiation at all let's change it. @AlexVanderbist please let me know about your decision.

AlexVanderbist commented on src/Config/FileConfigManager.php:
```diff
@@ -0,0 +1,142 @@
+        $this->path = $this->initPath($path);
+    }
+
+    private function initPath(string $path): string
```
For most OSS package we avoid `private` for all properties and functions. This allows the developers to be able to extend and easily override these in their own codebase.

```suggestion
    protected function initPath(string $path): string
```

kudashevs replied:
I haven't noticed that, my bad. Thank you for pointing it :)

AlexVanderbist commented on src/Config/FileConfigManager.php:
```diff
@@ -0,0 +1,142 @@
+
+        try {
+            file_put_contents($file, json_encode($options));
+        } catch (\Throwable) {
```
As discussed in laravel-ignition, explicitly importing `Throwable` is more in line with Laravel's codebase

AlexVanderbist commented on src/Config/FileConfigManager.php:
```diff
@@ -0,0 +1,142 @@
+        return rtrim($path, DIRECTORY_SEPARATOR);
+    }
+
+    public function createSource(): bool
```
`createSource` might be a bit vague. Maybe we can rename it to something a bit more explicite?

```suggestion
    public function createConfigFile(): bool
```

kudashevs replied:
I totally agree, that the term `source` is vague. However, I tried to make the contract as abstract as possible (in case if in the future somebody is going to use a database or other persistent source to keep the settings). If there is no need for the abstraction let's rename it according to your final decision about the `config` or `settings` file above. But if there is a chance that the persistent source will have changed in the future, let's try to come up with a less vague definition for that.

@AlexVanderbist please let me know about your decision.

AlexVanderbist commented on src/Config/FileConfigManager.php:
```diff
@@ -0,0 +1,142 @@
+        return rtrim($path, DIRECTORY_SEPARATOR);
+    }
+
+    public function createSource(): bool
```
Also, what do you think about calling this method from the `save()` method? We can still keep it public for use inside of tests tho.

kudashevs replied:
The idea to introduce and make `createSource()` method public was to give the possibility to create a settings storage somewhere outside. For example, if I don't want to use the `createSource` every time on the saving operation, I can just use it in a ServiceProvider to create it once. So, I'm still thinking about how to implement it and you are absolutely right that it's better to remove this call from `IgnitionConfig.php`.

I need to think about a better solution on how to create storage once and not use it on every save() call. Until then, I'll move it to the save() method.

AlexVanderbist commented on src/Config/IgnitionConfig.php:
```diff
@@ -29,6 +32,15 @@ public function setOption(string $name, string $value): self
 
+    private function initConfigManager(): ConfigManager
+    {
+        if (! function_exists('app')) {
+            return new FileConfigManager();
+        }
+
+        return app(ConfigManager::class);
```
Interesting approach! :D

kudashevs replied:
Frankly speaking, this approach is a little bit dangerous, because it leads to an exception if there is no bound instance in the container (according to PSR-11 and [according to the documentation](https://laravel.com/docs/9.x/container#psr-11) Laravel follows it). In my view this is the case where `fail fast` looks better than `fail soft`.

AlexVanderbist commented on src/Config/IgnitionConfig.php:
```diff
@@ -67,19 +69,9 @@ public function getConfigOptions(): array
-        } catch (Throwable) {
-            return false;
-        }
+        $this->manager->createSource();
```
Can this call be moved to inside of the `$manager->save()` method?

AlexVanderbist commented on src/Contracts/ConfigManager.php:
```diff
@@ -0,0 +1,17 @@
+
+interface ConfigManager
+{
+    /** @return array<string, mixed> */
+    public function load(): array;
+
+    /** @param array<string, mixed> $options */
+    public function save(array $options): bool;
+
+    /** @return array<string, mixed> */
+    public function getSource(): array;
```
I don't think these doc block typehints add a lot (I know we've used them in other places as well and I should really go over those and clean up 😅)

kudashevs replied:
I would rather leave them. At least, they show that it is expected to be associative arrays :)

AlexVanderbist commented on tests/Config/FileConfigManagerTest.php:
```diff
@@ -0,0 +1,100 @@
+    $this->assertArrayHasKey('test', $result);
+});
+
+// Helpers
```
```suggestion
```

AlexVanderbist reviewed (changes requested):
Great PR, I've commented on a couple of nitpicks but nothing major! Thanks for taking the time to add proper tests as well. Looking forward to merging this one 👍

kudashevs commented:
@AlexVanderbist I tried to push it gently but without any luck. Sorry for that :sob: I should've made another PR from another branch, probably. I don't know how to cope with such situations when the only option is to push with force. What are your suggestions, btw?

So, this is a modified solution reworked by taking into account suggestions and some new ideas:
-	Maybe we can rename this to `CONFIG_FILE_NAME` to keep it in line with the ConfigManager, etc...
Because I haven’t received any feedback on the explanations about my decision, I would like to stick to this naming for now. 
-	For most OSS package we avoid private for all properties and functions
I changed the visibility for all the methods which look worth being inherited. 
-	Explicitly importing `Throwable` is more in line with Laravel's codebase
Done for all the cases.
-	`createSource` might be a bit vague. Maybe we can rename it to something a bit more explicit?
The interface of the `ConfigManager` was change to:
```
  - load() 
  - save()
  - createPersistent()
  - getPersistentInfo()
```
that looks better to me and more self-explantory
-	Also, what do you think about calling this method from the save() method?
Done. The `FileConfigManager` uses a new flow. It creates a file at the initialization moment (`createFile()`) and retains two states ($path, $file) for the future use. That way it is not still necessary to create a persistent storage somewhere else.
-	Interesting approach!
The approach turned out a fig failure when I had tried to update `laravel-ignition` with this package. So, the solution was rethought and modified. Now, it should work without any possible issues.
-	Can this call be moved to inside of the $manager->save() method?
Done. It was mentioned above.
-	I don't think these doc block typehints add a lot
Because I haven’t received any feedback I left them as is.

Besides these refinements, I have finished a branch with [the integration of these changes](https://github.com/spatie/laravel-ignition/compare/main...kudashevs:settings_test) to the `laravel-ignition`. I haven't made a PR yet, because the branch uses a non-standard composer.json. Could you please give it a look?

As an afterthought, I would like to propose to extract the functionality based on user's home directory (everything under the initPathFromEnvironment()) to another class. It seems like these (use predefined filename vs use the dynamically retrieved user home directory) are two different responsibilities with different functionality (works per projects vs works for all the projects).

kudashevs commented:
I'll better close it.

AlexVanderbist commented:
Hey @kudashevs, sorry about the delay. I didn't mean to come across as disrespectful or uninterested. If you would still like to see this PR merged, feel free to re-open it and I'll be happy to continue looking at it.

kudashevs commented:
Hello @AlexVanderbist

No worries. I just thought that the proposition is not still interesting (because it turned out to be kind of messy, big, and complicated). Let's try it another time. I you don't mind I'll open a new PR.

AlexVanderbist commented:
Awesome, thanks! I'll go over it later today 👍
